src/ring_theory/localization.lean

```diff
@@ -170,71 +191,218 @@ by rw [mk_eq_mul_mk_one, mul_comm ↑s, coe_mul,
        rw [is_ring_hom.map_mul f, units.coe_mul, ← mul_assoc, ← mul_assoc];
        simp only [mul_right_comm],
   map_add := λ x y, localization.induction_on x $ λ r₁ s₁,
     localization.induction_on y $ λ r₂ s₂,
-    have g (s₁ * s₂) = g s₁ * g s₂, from units.ext (by simp only [hg, units.coe_mul]; exact is_ring_hom.map_mul f),
+    have g (s₁ * s₂) = g s₁ * g s₂,
+      from units.ext (by simp only [hg, units.coe_mul]; exact is_ring_hom.map_mul f),
     show _ * ↑(g (_ * _))⁻¹ = _ * _ + _ * _,
     by simp only [subtype.coe_mk, mul_one, one_mul, subtype.coe_eta, this, mul_inv_rev];
        simp only [is_ring_hom.map_mul f, is_ring_hom.map_add f, add_mul, (hg _).symm];
        simp only [mul_assoc, mul_comm, mul_left_comm, (units.coe_mul _ _).symm];
        rw [mul_inv_cancel_left, mul_left_comm, ← mul_assoc, mul_inv_cancel_right, add_comm] }
 
-@[reducible] def away (x : α) := loc α (powers x)
+noncomputable def lift (h : ∀ s ∈ S, is_unit (f s)) :
+  localization α S → β :=
+localization.lift' f (λ s, classical.some $ h s.1 s.2)
+  (λ s, by rw [← classical.some_spec (h s.1 s.2)]; refl)
+
+instance lift.is_ring_hom (h : ∀ s ∈ S, is_unit (f s)) :
+  is_ring_hom (lift f h) :=
+lift'.is_ring_hom _ _ _
+
+@[simp] lemma lift'_of (g : S → units β) (hg : ∀ s, (g s : β) = f s) (a : α) :
+  lift' f g hg (of a) = f a :=
+begin
+  delta lift',
+  erw quotient.lift_on_beta,
+  suffices : ((g 1)⁻¹ : units β) = 1,
+  { simp [this] },
+  { have := hg 1,
+    erw [is_ring_hom.map_one f] at this,
+    apply inv_inj,
+    apply units.ext,
+    simp [this], }
+end
+
+@[simp] lemma lift'_coe (g : S → units β) (hg : ∀ s, (g s : β) = f s) (a : α) :
+  lift' f g hg a = f a := lift'_of _ _ _ _
+
+@[simp] lemma lift_of (h : ∀ s ∈ S, is_unit (f s)) (a : α) :
+  lift f h (of a) = f a := lift'_of _ _ _ _
+
+@[simp] lemma lift_coe (h : ∀ s ∈ S, is_unit (f s)) (a : α) :
+  lift f h a = f a := lift'_of _ _ _ _
+
+@[simp] lemma lift'_circ_of (g : S → units β) (hg : ∀ s, (g s : β) = f s) :
+  lift' f g hg ∘ of = f := funext $ λ a, lift'_of _ _ _ a
+
+@[simp] lemma lift_circ_of (h : ∀ s ∈ S, is_unit (f s)) :
+  lift f h ∘ of = f := lift'_circ_of _ _ _
+
+@[simp] lemma lift'_of_comp (f : localization α S → β) [is_ring_hom f] :
+  lift' (f ∘ of) (units.map f ∘ to_units) (λ s, rfl) = f :=
+begin
+  apply funext,
+  rintros ⟨⟨r,s⟩⟩,
+  delta lift',
+  erw quotient.lift_on_beta,
+  change _ = f (mk _ _),
+  erw mk_eq,
+  rw is_ring_hom.map_mul f,
+  refl
+end
+
+@[simp] lemma lift_of_comp (f : localization α S → β) [is_ring_hom f] :
+  lift (f ∘ of) (λ s hs, is_unit_unit (units.map f (to_units ⟨s, hs⟩))) = f :=
+begin
+  convert lift'_of_comp f,
+  show lift' _ _ _ = lift' _ _ _,
+  congr' 1,
+  funext s,
+  ext,
+  rw classical.some_spec (is_unit_unit (units.map f (to_units s))),
+  refl,
+end
+
+protected lemma funext (f g : localization α S → β) [is_ring_hom f] [is_ring_hom g] (h : ∀ a : α, f a = g a) :
+  f = g :=
```

Comment:
How is this different from regular `funext`?

Reply:
Now you only need to check that `f` and `g` agree on the image of `of`. Which is typically a lot smaller than `univ`. (E.g. to check that to ring homs agree on `rat` it suffices to check that they agree on `int`.)